proposals/multi-field-value-classes.md

```diff
@@ -0,0 +1,691 @@
+
+Multi-field value classes are the next part of the Value classes feature after Inline classes.
+
+### Value classes
+
+Value classes are effective shallow-immutable data classes without identity, copy functions and component functions.
+
+Shallow immutability is chosen to make it possible to implement internally mutable classes with immutable public API such as `Lazy<T>`. However, there are domains where deep immutability is required, but allowing such requirement is currently outside the scope of the document.
+
+Value classes do not have `copy` methods due to their inconvenience for usage to mutate:
+* They are verbose and hard to understand when combined with constructions such as `if`s, loops.
+* They are unsuitable for nested structures: `a = a.copy(b = a.b.copy(c = f(a.b.c + 1)))` instead of `a.b.c = f(a.b.c)` for mutable classes.
+* They cannot be used for operators: `a = a.copy(b = a.b + 2)` instead of `a.b += 2` for mutable classes.
+
+Not yet implemented lens syntax is going to be used for convenient mutating mutable fields/variables of immutable object types. Other applications of the shallow-copying function `copy` do not exist for shallow-immutable value classes.
+
+Value classes do not have component functions as they are used for positional destructuring which is considered harmful because of exposing of the internal state and order of the fields. It will be replaced with not yet implemented named destructuring syntax.
+
+Customizing `equals` is not allowed until “typed equals” feature is released.
+
+### Further development: `VArray`s and reification
+
+One of other important steps of Value classes feature are Value arrays (`VArray`s) and reified classes and reified not-inline functions. They significantly extend applicability and optimizations of Value classes, **but they require MFVC to be already implemented** for that. Their importance is caused by frequent usage of containers that store and use stored values: they are currently handle value classes as boxed.
+
+`VArray` is the fundamental container and first reified class. `VArray`s solve the problem of effective arrays (with invariant type parameter) without necessity of declaring manual [`UIntArray`](https://github.com/JetBrains/kotlin/blob/30788566012c571aa1d3590912468d1ebe59983d/libraries/stdlib/unsigned/src/kotlin/UIntArray.kt#L15) for each type `UInt`. It also generifies existing `IntArray`, `LongArray`, `BooleanArray` etc. with scalable general `VArray<T>` which maps to `VArray<Int>`, `VArray<Long>`, `VArray<Boolean>` correspondingly.
+
+Having `VArray`s and reified functions also allows to write generic functions, extensions that operate on generic `VArray<T>`:
+```kotlin
+operator fun <reified T> VArray<T>.plus(other: VArray<T>) =
```

Comment:
`VArray`s are also not supported yet. It may be possible to lift up the limitation for Value Classes and primitives parameters. As there is no completed design for `VArray`s and reification, it is out of scope of the KEEP now.